app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt

```diff
@@ -490,16 +492,20 @@ class GleanMetricsService(private val context: Context) : MetricsService {
         // because it calls Google ad APIs that must be called *off* of the main thread.
         // These two things actually happen in parallel, but that should be ok because Glean
         // can handle events being recorded before it's initialized.
-        gleanInitializer = MainScope().launch {
+        gleanInitializer = MainScope().launch(start = CoroutineStart.LAZY) {
             Glean.registerPings(Pings)
         }
+
         // setStartupMetrics is not a fast function. It does not need to be done before we can consider
         // ourselves initialized. So, let's do it, well, later.
-        gleanSetStartupMetrics = MainScope().launch {
+        gleanSetStartupMetrics = MainScope().launch(start = CoroutineStart.LAZY) {
             setStartupMetrics()
         }
 
         context.settings().totalUriCount = 0
+
+        OnVisualCompletenessExecutionQueue.add(gleanInitializer)
+        OnVisualCompletenessExecutionQueue.add(gleanSetStartupMetrics)
```

Comment:
There is a hidden assumption here that we will always show UI and run through `HomeActivity`. But this is not true. A `FenixApplication` object will be created for every process that is launched by the app. We may launch the main process without UI (e.g. some background job runs) and we may launch other processes that have other UI (e.g. the crash reporter) and we may launch into the app and not go into the HomeActivity (not sure if those exist at the moment though). In those cases we would never run those methods. Is that a problem?

Reply:
> In those cases we would never run those methods. Is that a problem?

Yes - if the `registerPings` is not called, then Glean won't know about Fenix custom pings. If the metrics are not set, then they won't end up in any ping.

Reply:
> There is a hidden assumption here that we will always show UI and run through `HomeActivity`. But this is not true. A `FenixApplication` object will be created for every process that is launched by the app. We may launch the main process without UI (e.g. some background job runs) and we may launch other processes that have other UI (e.g. the crash reporter) and we may launch into the app and not go into the HomeActivity (not sure if those exist at the moment though). In those cases we would never run those methods. Is that a problem?

This is a very good point. I traced through the intent-handling code and it looks like any way that a user can launch the app (that's key to this statement) will go through the HomeActivity in some way. E.g., even when the app is cold started with an intent to view a URL it still goes through the HomeActivity.

That said, it never occurred to me that we would really want Glean metrics in the other cases you mentioned, @pocmo , so I appreciate you bringing them to my attention. I think that it would be great to have a way to distinguish the cases where there will be a UI loaded and where there will not be a UI loaded. In the former case we would delay these operations until visual completeness and the others we would do them immediately. I suppose it would be possible to draw this distinction by examining (the presence or lack of) intents during FenixApplication creation. What do you think, @pocmo ?